README.md
# Sourcebuster JS

#### 1.0.5 is merged to master

Read [upgrade guide](UPGRADING.md) for details.

## About
* Sourcebuster tracks the sources of your site’s visitors and stores the data in the cookies for further analysis.
* It handles sources-overriding just like Google Analytics does.
* It’s written in pure JavaScript, without any dependency from third-party libraries.
* You can use received data for:
 * phones change
 * site content change
 * export data within forms to your CRM or analysis system.

## Links

[Download](https://github.com/alexfedoseev/sourcebuster-js/archive/master.zip) &middot; [Upgrade guide](UPGRADING.md) &middot; [Changelog](CHANGELOG.md) &middot; [Test page](http://statica.alexfedoseev.com/sourcebuster-js/)

## Install

```python
# from `npm`
npm install --save sourcebuster

# or `bower`
bower install --save sourcebuster
```

Or you can [download this repo](https://github.com/alexfedoseev/sourcebuster-js/archive/master.zip) and use `sourcebuster.min.js` from `/dist` folder.

## Setup
I'm gonna walk you through the inline-HTML setup pattern, but keep in mind, that Sourcebuster is available as `CommonJS` / `AMD` module, so you can `require` it as dependency as well.

Script is written in pure JavaScript, doesn’t have any dependency on third-party libraries and doesn’t interacts with DOM’s objects, so it can be called as soon as you want it. Higher you’ll place it in the `<head>` tag, sooner you’ll get the cookies, whose data can be used for DOM’s objects manipulation (phones change etc.).

Place in the `<head>` tag:

```html
<script src="/path/to/sourcebuster.min.js"></script>
```

Or require it:

```javascript
var sbjs = require('sourcebuster');
```

It will expose Sourcebuster object — `sbjs`. This object has just 2 methods: `init` and `get`. Basically first one is `setter`, and the second one is `getter`.

So, lets initialize it:

```html
<script src="/path/to/sourcebuster.min.js"></script>
<script>

  sbjs.init();

</script>
```

This will:

  - set all the cookies
  - give you the data (available through the second method `sbjs.get`)

### Configure

`sbjs.init` accepts one argument: object with the settings. It's optional and usually it will be something like this:

```javascript
sbjs.init({
  domain: 'alexfedoseev.com',
  lifetime: 3,
  callback: doSmth
});
```

There are 11 types of user settings:
* lifetime
* session_length
* domain
* referrals
* organics
* typein_attributes
* timezone_offset
* campaign_param
* user_ip
* promocode
* callback

Now I'll show you the whole list of configurable options. The short description is provided in comments. The rest is explained below.

```javascript
sbjs.init({

  // Set custom expiration period for cookies in months
  // 6 months is default
  lifetime: 6,

  // Set custom session length in minutes
  // 30 minutes is default
  session_length: 30,

  // Set domain name in cookies
  domain: {
    host: 'alexfedoseev.com',
    isolate: false
  },

  // Set custom referral sources
  referrals: [
    {
      host: 't.co',          // This is host from Twitter's `http referer`
      medium: 'social',      // This is custom `utm_medium`, you can drop it and it'll be `referral`
      display: 'twitter.com' // And this is how you'll see it in the result data
    },
    {
      host: 'plus.url.google.com',
      display: 'plus.google.com'
    }
  ],

  // Set custom organic sources
  organics: [
    {
      host: 'bing.com',
      param: 'q',
      display: 'bing'
    }
  ],

  // Set `utm_source` & `utm_medium` values for `typein` traffic
  // Defaults are `(direct)` & `(none)`
  typein_attributes: {
    source: '(direct)',
    medium: '(none)'
  },

  // Set time zone
  timezone_offset: 3,

  // Set custom `utm_campaign` param
  campaign_param: 'my_adwords_campaign',

  // Set user ip
  user_ip: '192.168.1.1',

  // Set promocode
  // Default range is [100000..999999]
  promocode: {
    min: 100000,
    max: 999999
  },

  // Set callback-function,
  // It will be executed right after `sbjs` cookies will be set
  callback: doSmth
});
```

#### lifetime

```javascript
lifetime: 6
```

Setting up custom expiration period for `sbjs` cookies (in months). Default is 6.

#### session_length

```javascript
session_length: 30
```

Setting up user's session duration (in minutes). Default is 30.

This parameter affects only referral sources overriding. When the visitor comes on your site for the first time, we receive and store the data about his source. After some time this visitor can return to your website, but from another source, and we need to have some rules to decide — should we overwrite previous source or we should not.

The rules are the same with Google Analytics:
* UTM and Organic source always overrides any previous source.
* Typein never overrides a previous source.
* Referral source overrides previous source only if there is no user session at the moment. If it’s inside the same session — a referral source will never override previous source.

Explanation to `referral` logic: sometimes visitor within the current visit (session) comes to the website from the “source” which is not actually a “source”. For example, it can be visit from the email service, where he had a registration activation link.

#### domain

```javascript
domain: {
  host: 'alexfedoseev.com',
  isolate: false
}
```

Setting up cookies domain.

First of all lets talk about how script handles **the absence** of this custom option: if no value is set, cookies will be placed for **current domain and all its subdomains**.

Why it is so. Lets say the site doesn't have subdomains to share traffic with, and now it doesn't matter: do we share the cookies with subdomains or we do not. But what's gonna happen if someday subdomains will occur? If cookies wasn't shared — subdomains won't get them, it means they won't share the traffic, and visitors, which came from main domain to subdomains (and vise versa), will be considered as `referral` traffic.

So by default the cookies are shared. However if you don't want to share them — use `isolate: true`, to isolate domain.

Let’s take a look at further examples.

**No. 1**  
The first scenario. You have a site: **wow.com**. Also you have a blog on this site: **blog.wow.com**. And you don’t want to separate traffic between them. It means that when the visitor come from **blog.wow.com** to **wow.com**, it won’t be a new visit for **wow.com** (with **blog.wow.com** as Referral source), it will be a common inner click without source change (to sourcebuster.js this scenario will be equal to inner page change: from *wow.com/about* to *wow.com/contacts* for example). To achive this you have to add this line on both sites: **wow.com & blog.wow.com**.

```javascript
domain: 'wow.com'
```

**No. 2**  
That was simple. Now lets take a look at opposite scenario — you want to separate traffic between your subdomains and consider it as Referral. There is the main site — **wow.com**, and there is a blog on this site — **blog.wow.com**, which also has users subdomains — **alex.blog.wow.com** is one of them. You don’t want to separate traffic between **blog.wow.com** and **alex.blog.wow.com**, but you do want to separate it between all the blogs and the main site. Here’s how we can sort it out:

```javascript
// put this on the main domain wow.com
domain: {
  host: 'wow.com',
  isolate: true
}

// and this on the blogs subdomains (blog.wow.com & alex.blog.wow.com) 
domain: 'blog.wow.com'
```

Pay attention to the `isolate: true` param in setting for the main domain. Use it only when all incoming traffic from all subdomains should be referral in relation to provided host.

In our example if the visitor came for the first time on the main site **wow.com** by clicking on the link in user’s blog **alex.blog.wow.com**, his source (for the **wow.com**) will be **alex.blog.wow.com** (traffic type: *referral*).


CHANGELOG.md
## 1.0.5

Features:

  - added `t.co` (`host: 't.co', display: 'twitter.com'`) and `plus.url.google.com` (`host: 'plus.url.google.com', display: 'plus.google.com'`) to default `referral` sources. You still can override it by your custom setting (to mark it as `social` for example).

## 1.0.4

Fixes:

  - fixed IE8 bug

## 1.0.3

Fixes:

  - fixed month in dates (`sbjs.get.first_add.fd` & `sbjs.get.current_add.fd`): it was -1 from actual one

## 1.0.2

Fixes:

  - fixed validation of `lifetime`, `session_length` and `timezone_offset` params

## 1.0.0

[Upgrade guide](UPGRADING.md) &middot; [Updated documentation](README.md)

Changes:

_Major_

  - API. Rewritten from the ground. See [upgrade guide](UPGRADING.md) and [updated documentation](README.md).
  - Changed cookie's domain handling — also see [upgrade guide](UPGRADING.md) and [updated documentation](README.md).
  - Changed default `utm_source` & `utm_medium` values for `typein` traffic — you know [what to do](UPGRADING.md) :)

_Minor_

  - Changed default timezone from UTC to user's system (default — means no custom value provided during `sbjs` initialization).

_Under the hood_

  - In short: `mess` became `structure`. Check the sources if interested.
  - Added migration tool, which updates visitor's cookies to `1.x.x`.
  - Changed params delimiter in cookies: from `|` to `|||`. Cookies will be automatically upgraded by migration tool without data loss. You don't have to worry about this change unless you parse these cookies by yourself.

Features:

  - Added `callback` option to set custom function, which will be executed right after `sbjs` cookies will be set. This function will get `sbjs.get`-object as argument (it contains all detected `sbjs` data).
  - Added ability to set custom cookies expiration period
  - Added `term` detection from Yandex Direct traffic (if there is no `utm_term` in URL)
  - Added `visits` counter: available through `sbjs.get.udata.vst`
  - Added `pages` counter for current session: available through `sbjs.get.session.pgs`
  - Added `current page` url for current session: available through `sbjs.get.session.cpg`
  - And Sourcebuster `1.0.0` available as `CommonJS` / `AMD` module.

## 0.0.7

Features:

  - add `sbjs_current_add` cookie, which holds same data as `sbjs_first_add` cookie, but for the visit when current source was written
  - add info about first referer (see changes below)

Changes:

  - move current `referer` data from stand-alone `sbjs_referer` cookie to `sbjs_current_add` cookie (and also to `sbjs_first_add` cookie, now you can get first and current referers via `get_sbjs.first_add.rf` & `get_sbjs.current_add.rf` getters)

## 0.0.6

Features:

  - add `_setPromocode` user setting to set promocode for visitor;

## 0.0.5

Features:

  - add `_setTypeinAttributes` user setting to set custom `source` and `medium` for `typein` traffic
  - set value of `content` = `referer_path` in cookies for `referral` traffic

## 0.0.4

Features:

  - add predefined organic sources into the core (see comments and the list of added sources [in the docs](./README.md#_addorganicsource))
  - add Sourcebuster to `bower` ([bower.io](http://bower.io)). Now you can [install it via... `bower`](./README.md#install)

Changes:

  - switch from Closure Compiler to UglifyJS minification library
  - switch from deprecated `escape()`/`unescape()` methods to custom `encode_data()`/`decode_data()` functions (which uses `encodeURIComponent()`/`decodeURIComponent()` with some additions) to set/get cookies. **Please pay attention to this if you parse `sbjs` cookies by yourself.**

## 0.0.3

Features:

  - add `_setTimeZoneOffset` user setting to set default time zone
  - add alias param to custom organic source (added via `_addOrganicSource`)

Changes:

  - change simple and advanced setup methods
  - change datetime format to `yyyy-mm-dd hh:mm:ss` (`2014-06-11 17:28:26`)
  - store datetime in UTC by default

js/sourcebuster.js
// It's Sourcebuster [JS Edition], baby! (poolparty)
// Author: Alex Fedoseev (www.alexfedoseev.com)
// Version: 0.0.7

// Github: https://github.com/alexfedoseev/sourcebuster-js
// Blog post (rus): http://www.alexfedoseev.com/post/40/sourcebuster-js

// Ready Steady
// Let's set some vars first

// SBJS cookies
var SBJS_CURRENT_COOKIE = 'sbjs_current',
    SBJS_CURRENT_ADD_COOKIE = 'sbjs_current_add',
    SBJS_FIRST_COOKIE = 'sbjs_first',
    SBJS_FIRST_ADD_COOKIE = 'sbjs_first_add',
    SBJS_SESSION_COOKIE = 'sbjs_session',
    SBJS_UDATA_COOKIE = 'sbjs_udata',
    SBJS_PROMOCODE_COOKIE = 'sbjs_promo',
    SBJS_COOKIE_EXPIRES = 1000000;

// Encode/Decode helpers
function encode_data(s) {
  return encodeURIComponent(s).replace(/\!/g, "%21")
                              //.replace(/\-/g, "%2D")
                              //.replace(/\_/g, "%5F")
                              //.replace(/\./g, "%2E")
                              .replace(/\~/g, "%7E")
                              .replace(/\*/g, "%2A")
                              .replace(/\'/g, "%27")
                              .replace(/\(/g, "%28")
                              .replace(/\)/g, "%29");
}

function decode_data(s) {
  try {
    return decodeURIComponent(s).replace(/\%21/g, "!")
                                //.replace(/\%2D/g, "-")
                                //.replace(/\%5F/g, "_")
                                //.replace(/\%2E/g, ".")
                                .replace(/\%7E/g, "~")
                                .replace(/\%2A/g, "*")
                                .replace(/\%27/g, "'")
                                .replace(/\%28/g, "(")
                                .replace(/\%29/g, ")");
  } catch(err1) {
    // try unescape for backward compatibility
    try { return unescape(s); } catch(err2) { return ""; }
  }
}

// It's not a Rails, we don't have a waiters here
function set_cookie(name, value, minutes, domain, excl_subdomains) {
  var expires, basehost;

  if (minutes) {
    var date = new Date();
    date.setTime(date.getTime() + (minutes * 60 * 1000));
    expires = '; expires=' + date.toGMTString();
  } else {
    expires = '';
  }
  if (domain && !excl_subdomains) {
    basehost = ';domain=.' + domain;
  } else {
    basehost = '';
  }
  document.cookie = encode_data(name) + '=' + encode_data(value) + expires + basehost + '; path=/';
}

function get_cookie(name) {
  var nameEQ = encode_data(name) + '=',
      ca = document.cookie.split(';');

  for (var i = 0; i < ca.length; i++) {
    var c = ca[i];
    while (c.charAt(0) === ' ') { c = c.substring(1, c.length); }
    if (c.indexOf(nameEQ) === 0) { return decode_data(c.substring(nameEQ.length, c.length)); }
  }
  return null;
}

function destroy_cookie(name) {
  set_cookie(name, '', -1);
}

// Set predefined organic custom sources
var _sbjs = _sbjs || [];
_sbjs.push(['_addOrganicSource', 'bing.com', 'q', 'bing']);
_sbjs.push(['_addOrganicSource', 'yahoo.com', 'p', 'yahoo']);
_sbjs.push(['_addOrganicSource', 'about.com', 'q', 'about']);
_sbjs.push(['_addOrganicSource', 'aol.com', 'q', 'aol']);
_sbjs.push(['_addOrganicSource', 'ask.com', 'q', 'ask']);
_sbjs.push(['_addOrganicSource', 'globososo.com', 'q', 'globo']);
_sbjs.push(['_addOrganicSource', 'go.mail.ru', 'q', 'go.mail.ru']);
_sbjs.push(['_addOrganicSource', 'rambler.ru', 'query', 'rambler']);
_sbjs.push(['_addOrganicSource', 'tut.by', 'query', 'tut.by']);

(function sourcebuster_js() {

      // Traffic types
  var SBJS_UTM = 'utm',
      SBJS_ORGANIC = 'organic',
      SBJS_REFERRAL = 'referral',
      SBJS_TYPEIN = 'typein';

      // Medium types
  var SBJS_REFERER_REFERRAL = 'referral',
      SBJS_REFERER_ORGANIC = 'organic',
      SBJS_REFERER_SOCIAL = 'social';

      // Aliases for params names in main cookies
  var SBJS_TYPE_ALIAS = 'typ',
      SBJS_SOURCE_ALIAS = 'src',
      SBJS_MEDIUM_ALIAS = 'mdm',
      SBJS_CAMPAIGN_ALIAS = 'cmp',
      SBJS_CONTENT_ALIAS = 'cnt',
      SBJS_TERM_ALIAS = 'trm';

  var SBJS_FIRST_DATE_ALIAS = 'fd',
      SBJS_ENTRANCE_POINT_ALIAS = 'ep',
      SBJS_REFERRAL_URL_ALIAS = 'rf';

  var SBJS_USER_IP_ALIAS = 'uip',
      SBJS_USER_AGENT_ALIAS = 'uag';

  var SBJS_PROMOCODE_ALIAS = 'code';

  var SBJS_NONE = '(none)',
      SBJS_OOPS = '(Houston, we have a problem)';

      // source params
  var __sbjs_type,
      __sbjs_source,
      __sbjs_medium,
      __sbjs_campaign,
      __sbjs_content,
      __sbjs_term;


  // Set user params
  for (var i = 0; i < _sbjs.length; i++) {

    if (_sbjs[i][0] === '_setBaseHost') {
      var SBJS_BASEHOST = _sbjs[i][1];
      var SBJS_IS_TRUE_BASEHOST;
      if (_sbjs[i].length > 2) {
        SBJS_IS_TRUE_BASEHOST = _sbjs[i][2];
      } else {
        SBJS_IS_TRUE_BASEHOST = true;
      }
    }

    if (_sbjs[i][0] === '_setSessionLength') {
      var SBJS_SESSION_LENGTH = parseInt(_sbjs[i][1]);
    }

    if (_sbjs[i][0] === '_setUserIP') {
      var SBJS_USER_IP = _sbjs[i][1];
    }

    if (_sbjs[i][0] === '_addOrganicSource') {
      var SBJS_CUSTOM_SOURCES_ORGANIC = SBJS_CUSTOM_SOURCES_ORGANIC || [];
      SBJS_CUSTOM_SOURCES_ORGANIC.push(_sbjs[i]);
    }

    if (_sbjs[i][0] === '_addReferralSource') {
      var SBJS_CUSTOM_SOURCES_REFERRAL = SBJS_CUSTOM_SOURCES_REFERRAL || [];
      SBJS_CUSTOM_SOURCES_REFERRAL.push(_sbjs[i]);
    }

    if (_sbjs[i][0] === '_setTypeinAttributes') {
      var SBJS_TYPEIN_CUSTOM_SOURCE = _sbjs[i][1] || SBJS_TYPEIN;
      var SBJS_TYPEIN_CUSTOM_MEDIUM = _sbjs[i][2] || SBJS_TYPEIN;
    }

    if (_sbjs[i][0] === '_setCampaignParam') {
      var SBJS_CAMPAIGN_PARAM = _sbjs[i][1];
    }

    if (_sbjs[i][0] === '_setTimeZoneOffset') {
      var SBJS_TIMEZONE_OFFSET = parseInt(_sbjs[i][1]);
    }

    if (_sbjs[i][0] === '_setPromocode') {
      var SBJS_SET_PROMO = true,
          SBJS_PROMOCODE_MIN = parseInt(_sbjs[i][1]) || 100000,
          SBJS_PROMOCODE_MAX = parseInt(_sbjs[i][2]) || 999999;
    }

  }

  // A few more helpers
  var get_param = function () {
    var query_string = {},
        query = window.location.search.substring(1),
        vars = query.split('&');

    for (var i = 0; i < vars.length; i++) {
      var pair = vars[i].split('=');
      if (typeof query_string[pair[0]] === 'undefined') {
        query_string[pair[0]] = pair[1];
      } else if (typeof query_string[pair[0]] === 'string') {
        var arr = [ query_string[pair[0]], pair[1] ];
        query_string[pair[0]] = arr;
      } else {
        query_string[pair[0]].push(pair[1]);
      }
    }
    return query_string;
  }();


  // parseUri 1.2.2
  // (c) Steven Levithan <stevenlevithan.com>
  // MIT License

  function parseUri(str) {
    var o = parseUri.options,
        m = o.parser[o.strictMode ? 'strict' : 'loose'].exec(str),
        uri = {},
        i = 14;

    while (i--) { uri[o.key[i]] = m[i] || ''; }

    uri[o.q.name] = {};
    uri[o.key[12]].replace(o.q.parser, function ($0, $1, $2) {
      if ($1) { uri[o.q.name][$1] = $2; }
    });

    return uri;
  }

  parseUri.options = {
    strictMode: false,
    key: ['source','protocol','authority','userInfo','user','password','host','port','relative','path','directory','file','query','anchor'],
    q: {
      name:   'queryKey',
      parser: /(?:^|&)([^&=]*)=?([^&]*)/g
    },
    parser: {
      strict: /^(?:([^:\/?#]+):)?(?:\/\/((?:(([^:@]*)(?::([^:@]*))?)?@)?([^:\/?#]*)(?::(\d*))?))?((((?:[^?#\/]*\/)*)([^?#]*))(?:\?([^#]*))?(?:#(.*))?)/,
      loose:  /^(?:(?![^:@]+:[^:@\/]*@)([^:\/?#.]+):)?(?:\/\/)?((?:(([^:@]*)(?::([^:@]*))?)?@)?([^:\/?#]*)(?::(\d*))?)(((\/(?:[^?#](?![^?#\/]*\.[^?#\/.]+(?:[?#]|$)))*\/?)?([^?#\/]*))(?:\?([^#]*))?(?:#(.*))?)/
    }
  };
  // parseUri 1.2.2


  function escape_regexp(string) {
    return string.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
  }


  // Let's get this party started
  function set_first_and_current_add_cookie() {
    var basehost = typeof SBJS_BASEHOST !== 'undefined' ? SBJS_BASEHOST : undefined;
    var is_true_basehost = typeof SBJS_IS_TRUE_BASEHOST !== 'undefined' ? SBJS_IS_TRUE_BASEHOST : true;

    set_cookie(SBJS_CURRENT_ADD_COOKIE, combine_sbjs_source_add_data_string(), SBJS_COOKIE_EXPIRES, basehost, !is_true_basehost);
    if (!get_cookie(SBJS_FIRST_ADD_COOKIE)) {
      set_cookie(SBJS_FIRST_ADD_COOKIE, combine_sbjs_source_add_data_string(), SBJS_COOKIE_EXPIRES, basehost, !is_true_basehost);
    }
  }

  function main_cookie_data() {
    var data;
    if (typeof get_param.utm_source !== 'undefined' || 
        typeof get_param.utm_medium !== 'undefined' ||
        typeof get_param.utm_campaign !== 'undefined' ||
        typeof get_param.utm_content !== 'undefined' ||
        typeof get_param.utm_term !== 'undefined' ||
        typeof get_param.gclid !== 'undefined' ||
        typeof get_param[SBJS_CAMPAIGN_PARAM] !== 'undefined') {
      set_first_and_current_add_cookie();
      data = get_data(SBJS_UTM);
    } else if (check_referer(SBJS_ORGANIC)) {
      set_first_and_current_add_cookie();
      data = get_data(SBJS_ORGANIC);
    } else if (!get_cookie(SBJS_SESSION_COOKIE) && check_referer(SBJS_REFERRAL)) {
      set_first_and_current_add_cookie();
      data = get_data(SBJS_REFERRAL);
    } else if (!get_cookie(SBJS_FIRST_COOKIE) && !get_cookie(SBJS_CURRENT_COOKIE)) {
      set_first_and_current_add_cookie();
      data = get_data(SBJS_TYPEIN);
    } else {
      return get_cookie(SBJS_CURRENT_COOKIE);
    }
    return data;
  }

  function get_data(type) {
    switch (type) {
      case SBJS_UTM:
        __sbjs_type = SBJS_UTM;

        if (typeof get_param.utm_source !== 'undefined') {
          __sbjs_source = get_param.utm_source;
        } else if (typeof get_param.gclid !== 'undefined') {
          __sbjs_source = 'google';
        } else {
          __sbjs_source = SBJS_NONE;
        }
        
        if (typeof get_param.utm_medium !== 'undefined') {
          __sbjs_medium = get_param.utm_medium;
        } else if (typeof get_param.gclid !== 'undefined') {
          __sbjs_medium = 'cpc';
        } else {
          __sbjs_medium = SBJS_NONE;
        }

        if (typeof get_param.utm_campaign !== 'undefined') {
          __sbjs_campaign = get_param.utm_campaign;
        } else if (typeof get_param[SBJS_CAMPAIGN_PARAM] !== 'undefined') {
          __sbjs_campaign = get_param[SBJS_CAMPAIGN_PARAM];
        } else if (typeof get_param.gclid !== 'undefined') {
          __sbjs_campaign = 'google_cpc';
        } else {
          __sbjs_campaign = SBJS_NONE;
        }

        __sbjs_content = get_param.utm_content || SBJS_NONE;
        __sbjs_term = get_param.utm_term || SBJS_NONE;
        break;

      case SBJS_ORGANIC:
        __sbjs_type = SBJS_ORGANIC;
        __sbjs_source = __sbjs_source || clean_host(document.referrer);
        __sbjs_medium = SBJS_REFERER_ORGANIC;
        __sbjs_campaign = SBJS_NONE;
        __sbjs_content = SBJS_NONE;
        __sbjs_term = SBJS_NONE;
        break;

      case SBJS_REFERRAL:
        __sbjs_type = SBJS_REFERRAL;
        __sbjs_source = __sbjs_source || clean_host(document.referrer);
        __sbjs_medium = __sbjs_medium || SBJS_REFERER_REFERRAL;
        __sbjs_campaign = SBJS_NONE;
        __sbjs_content = parseUri(document.referrer).path;
        __sbjs_term = SBJS_NONE;
        break;

      case SBJS_TYPEIN:
        __sbjs_type = SBJS_TYPEIN;
        __sbjs_source = SBJS_TYPEIN_CUSTOM_SOURCE || SBJS_TYPEIN;
        __sbjs_medium = SBJS_TYPEIN_CUSTOM_MEDIUM || SBJS_TYPEIN;
        __sbjs_campaign = SBJS_NONE;
        __sbjs_content = SBJS_NONE;
        __sbjs_term = SBJS_NONE;
        break;

      default:
        __sbjs_type = SBJS_OOPS;
        __sbjs_source = SBJS_OOPS;
        __sbjs_medium = SBJS_OOPS;
        __sbjs_campaign = SBJS_OOPS;
        __sbjs_content = SBJS_OOPS;
        __sbjs_term = SBJS_OOPS;
    }
    var data = {  sbjs_type: __sbjs_type,
                  sbjs_source: __sbjs_source,
                  sbjs_medium: __sbjs_medium,
                  sbjs_campaign: __sbjs_campaign,
                  sbjs_content: __sbjs_content,
                  sbjs_term: __sbjs_term
        };
    return combine_sbjs_main_data_string(data);
  }

  function clean_host(request) {
    return parseUri(request).host.replace('www.', '');
  }

  function check_referer(type) {
    var referer = document.referrer;
    switch(type) {
      case SBJS_ORGANIC:
        return (!!referer && check_referer_host(referer) && is_organic(referer));
      case SBJS_REFERRAL:
        return (!!referer && check_referer_host(referer) && is_referral(referer));
      default:
        return false;
    }
  }

  function check_referer_host(referer) {
    if (typeof SBJS_BASEHOST !== 'undefined' && SBJS_BASEHOST.length > 0) { 
      if (SBJS_IS_TRUE_BASEHOST) {
        var host_regex = new RegExp('^(.*\\.)?' + escape_regexp(SBJS_BASEHOST) + '$', 'i');
        return !(!!clean_host(referer).match(host_regex));
      } else {
        return (clean_host(referer) !== clean_host(SBJS_BASEHOST));
      }
    } else { 
      return (clean_host(referer) !== clean_host(location.href));
    }
  }

  function is_organic(referer) {
    var y_host = 'yandex',
        y_param = 'text',
        g_host = 'google';
    var y_host_regex = new RegExp('^(.*\\.)?' + escape_regexp(y_host) + '\\..{2,9}$'),
        y_param_regex = new RegExp('.*' + escape_regexp(y_param) + '=.*'),
        g_host_regex = new RegExp('^(www\\.)?' + escape_regexp(g_host) + '\\..{2,9}$');

    if (!!parseUri(referer).query && 
        !!parseUri(referer).host.match(y_host_regex) && 
        !!parseUri(referer).query.match(y_param_regex)) {
      __sbjs_source = y_host;
      return true;
    } else if (!!parseUri(referer).host.match(g_host_regex)) {
      __sbjs_source = g_host;
      return true;
    } else if (!!parseUri(referer).query && typeof SBJS_CUSTOM_SOURCES_ORGANIC !== 'undefined' && SBJS_CUSTOM_SOURCES_ORGANIC.length > 0) {
      for (var i = 0; i < SBJS_CUSTOM_SOURCES_ORGANIC.length; i++) {
        if (SBJS_CUSTOM_SOURCES_ORGANIC[i].length >= 3 && parseUri(referer).host.match(new RegExp('^(.*\\.)?' + escape_regexp(SBJS_CUSTOM_SOURCES_ORGANIC[i][1]) + '$', 'i')) && parseUri(referer).query.match(new RegExp('.*' + escape_regexp(SBJS_CUSTOM_SOURCES_ORGANIC[i][2]) + '=.*', 'i'))) {
          __sbjs_source = SBJS_CUSTOM_SOURCES_ORGANIC[i][3] || SBJS_CUSTOM_SOURCES_ORGANIC[i][1];
          return true;
        }
        if (i + 1 === SBJS_CUSTOM_SOURCES_ORGANIC.length) {
          return false;
        }
      }
    } else {
      return false;
    }
  }

  function is_referral(referer) {
    if (typeof SBJS_CUSTOM_SOURCES_REFERRAL !== 'undefined' && SBJS_CUSTOM_SOURCES_REFERRAL.length > 0) {
      for (var i = 0; i < SBJS_CUSTOM_SOURCES_REFERRAL.length; i++) {
        if (SBJS_CUSTOM_SOURCES_REFERRAL[i].length > 1 && parseUri(referer).host.match(new RegExp('^(.*\\.)?' + escape_regexp(SBJS_CUSTOM_SOURCES_REFERRAL[i][1]) + '$', 'i'))) {
          __sbjs_source = SBJS_CUSTOM_SOURCES_REFERRAL[i][3] || SBJS_CUSTOM_SOURCES_REFERRAL[i][1];
          __sbjs_medium = SBJS_CUSTOM_SOURCES_REFERRAL[i][2] || SBJS_REFERER_REFERRAL;
          return true;
        }
        if (i + 1 === SBJS_CUSTOM_SOURCES_REFERRAL.length) {
          __sbjs_source = clean_host(referer);
          return true;
        }
      }
    } else {
      __sbjs_source = clean_host(referer);
      return true;
    }
  }

  function set_leading_zero_to_int(num, size) {
    var s = num + '';
    while (s.length < size) { s = '0' + s; }
    return s;
  }

  function set_date(date) {
    var utc_offset = date.getTimezoneOffset() / 60,
        now_hours = date.getHours(),
        custom_offset;
    if (typeof SBJS_TIMEZONE_OFFSET !== 'undefined' && typeof SBJS_TIMEZONE_OFFSET === 'number' && SBJS_TIMEZONE_OFFSET % 1 === 0) {
      custom_offset = SBJS_TIMEZONE_OFFSET;
    } else {
      custom_offset = 0;
    }
    date.setHours(now_hours + utc_offset + custom_offset);

    var date_string,
        year = date.getFullYear(),
        month = set_leading_zero_to_int(date.getMonth(), 2),
        day = set_leading_zero_to_int(date.getDate(), 2),
        hour = set_leading_zero_to_int(date.getHours(), 2),
        minute = set_leading_zero_to_int(date.getMinutes(), 2),
        second = set_leading_zero_to_int(date.getSeconds(), 2);

    date_string = year + '-' + month + '-' + day + ' ' + hour + ':' + minute + ':' + second;
    return date_string;
  }

  function random_int(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }

  // Pre-pack data
  function combine_sbjs_main_data_string(data) {
    return  SBJS_TYPE_ALIAS + '=' + data.sbjs_type + '|' + 
            SBJS_SOURCE_ALIAS + '=' + data.sbjs_source + '|' + 
            SBJS_MEDIUM_ALIAS + '=' + data.sbjs_medium + '|' + 
            SBJS_CAMPAIGN_ALIAS + '=' + data.sbjs_campaign + '|' + 
            SBJS_CONTENT_ALIAS + '=' + data.sbjs_content + '|' + 
            SBJS_TERM_ALIAS + '=' + data.sbjs_term;
  }

  function combine_sbjs_user_data_string() {
    var user_agent_string = '|' + SBJS_USER_AGENT_ALIAS + '=' + navigator.userAgent;
    if (typeof SBJS_USER_IP !== 'undefined' && SBJS_USER_IP.length > 0) {
      return (SBJS_USER_IP_ALIAS + '=' + SBJS_USER_IP + user_agent_string);
    } else {
      return (SBJS_USER_IP_ALIAS + '=' + SBJS_NONE + user_agent_string);
    }
  }

  function combine_sbjs_source_add_data_string() {
    var current_date = new Date();
    var referer = document.referrer || SBJS_NONE;
    return (SBJS_FIRST_DATE_ALIAS + '=' + set_date(current_date) + '|' + SBJS_ENTRANCE_POINT_ALIAS + '=' + location.href + '|' + SBJS_REFERRAL_URL_ALIAS + '=' + referer);
  }

  function combine_sbjs_promocode_string() {
    return (SBJS_PROMOCODE_ALIAS + '=' + set_leading_zero_to_int(random_int(SBJS_PROMOCODE_MIN, SBJS_PROMOCODE_MAX), SBJS_PROMOCODE_MAX.toString().length));
  }

  function set_sbjs_data() {

    var session_length = typeof SBJS_SESSION_LENGTH !== 'undefined' && SBJS_SESSION_LENGTH > 0 ? SBJS_SESSION_LENGTH : 30;
    var basehost = typeof SBJS_BASEHOST !== 'undefined' ? SBJS_BASEHOST : undefined;
    var is_true_basehost = typeof SBJS_IS_TRUE_BASEHOST !== 'undefined' ? SBJS_IS_TRUE_BASEHOST : true;
    var session_first_pageload = !get_cookie(SBJS_SESSION_COOKIE) ? '0' : '1';
    
    set_cookie(SBJS_CURRENT_COOKIE, main_cookie_data(), SBJS_COOKIE_EXPIRES, basehost, !is_true_basehost);
    if (!get_cookie(SBJS_FIRST_COOKIE)) {
      set_cookie(SBJS_FIRST_COOKIE, get_cookie(SBJS_CURRENT_COOKIE), SBJS_COOKIE_EXPIRES, basehost, !is_true_basehost);
    }
    set_cookie(SBJS_SESSION_COOKIE, session_first_pageload, session_length, basehost, !is_true_basehost);
    set_cookie(SBJS_UDATA_COOKIE, combine_sbjs_user_data_string(), SBJS_COOKIE_EXPIRES, basehost, !is_true_basehost);

    if (SBJS_SET_PROMO && !get_cookie(SBJS_PROMOCODE_COOKIE)) {
      set_cookie(SBJS_PROMOCODE_COOKIE, combine_sbjs_promocode_string(), SBJS_COOKIE_EXPIRES, basehost, !is_true_basehost);
    }

  }

  // Boom!
  set_sbjs_data();

}).call(this);


// What we've got here is getter to communicate
var get_sbjs = function() {
  var cookies = {},
      cookies_names_src = [
        SBJS_CURRENT_COOKIE,
        SBJS_CURRENT_ADD_COOKIE,
        SBJS_FIRST_COOKIE,
        SBJS_FIRST_ADD_COOKIE,
        SBJS_UDATA_COOKIE,
        SBJS_PROMOCODE_COOKIE
      ];

  function unsbjs(string) {
    return string.replace('sbjs_', '');
  }

  for (var i1 = 0; i1 < cookies_names_src.length; i1++) {
    var cookie_array;
    cookies[unsbjs(cookies_names_src[i1])] = {};
    if (get_cookie(cookies_names_src[i1])) {
      cookie_array = get_cookie(cookies_names_src[i1]).split('|');
    } else {
      cookie_array = [];
    }
    for (var i2 = 0; i2 < cookie_array.length; i2++) {
      var tmp_array = cookie_array[i2].split('='),
          result_array = tmp_array.splice(0, 1);
      result_array.push(tmp_array.join('='));
      cookies[unsbjs(cookies_names_src[i1])][result_array[0]] = decode_data(result_array[1]);
    }
  }
  return cookies;
}();
